### DISCUSSION TITLE
MethodGroup causes ambiguity over the same lambda expression

### DISCUSSION THREAD

**Discussion by JackHerring** [dotnet/csharplang#1245] [General]
Given the following interface:

```
public interface IBuilder
{
    void Property( Func<ValueTask<string>> source );
    void Property( Func<Task<string>> source );
}
```

The following does not compile:

```
class Program
{
    static void Main( string[] args )
    {
        IBuilder builder = null;

        builder.Property( Source );
    }

    private static ValueTask<string> Source()
    {
        return new ValueTask<string>();
    }
}
```
![image](https://user-images.githubusercontent.com/8951081/34768069-bd1379bc-f5f1-11e7-9645-42ef46d52979.png)

**However**, if you pass a lambda expression which just invokes `Source` then it compiles fine:

`builder.Property( () => Source() );`

Why is the first call (using method grouping) failing? A bug..?

**Comment by JackHerring** [comment]
FYI have raised an issue @jrote1

**Comment by DavidArno** [comment]
This appears unrelated to `Task` and `ValueTask` and will occur for any two types, eg:
```cs
public interface IBuilder
{
    void Property( Func<string> source );
    void Property( Func<int> source );
}

class Program
{
    static void Main( string[] args )
    {
        IBuilder builder = null;

        builder.Property( Source ); // ambiguous call
    }

    private static int Source()
    {
        return 0;
    }
}
```

Method groups are resolved using the arity and types of the parameters, not the return type. So presumably since you are using `Func<T>`, and thus there are no parameters, method group resolution can't be applied here?

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/98

> For a method group conversion, candidate methods whose return type doesn't match up with the delegate's return type are removed from the set.
